chasm/search_attribute.go

```diff
@@ -0,0 +1,340 @@
+	"time"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+// CHASM Search Attribute User Guide:
+//
+// Below contains exported CHASM search attribute field constants. These predefined fields correspond to the exact column name in Visibility storage.
+// For each root component, search attributes can be mapped from a user defined alias to these fields.
+//
+// To define a CHASM search attribute, create this as a package/global scoped variable. Below is an example:
+// var testComponentCompletedSearchAttribute = NewSearchAttributeBool("Completed", SearchAttributeFieldBool01)
+// var testComponentFailedSearchAttribute = NewSearchAttributeBool("Failed", SearchAttributeFieldBool02)
+// var testComponentStartTimeSearchAttribute = NewSearchAttributeTime("StartTime", SearchAttributeFieldDateTime01)
+//
+// Each CHASM search attribute field is associated with a specific indexed value type. The Value() method of a search attribute
+// specifies the supported value to set at compile time. eg. DateTime values must be set with a time.Time typed value.
+//
+// Each root component can ONLY use a predefined search attribute field ONCE. Developers should NOT reassign aliases to different fields.
+// Reassiging fields to different aliases is a breaking change during visibility queries.
+//
+// To register these search attributes with the CHASM Registry, use the WithSearchAttributes() option when creating the component in the library.
+// eg.
+// NewRegistrableComponent[T]("testcomponent", WithSearchAttributes(testComponentCompletedSearchAttribute, testComponentStartTimeSearchAttribute))
+var (
+	SearchAttributeFieldBool01 = newSearchAttributeFieldBool(1)
+	SearchAttributeFieldBool02 = newSearchAttributeFieldBool(2)
+
+	SearchAttributeFieldDateTime01 = newSearchAttributeFieldDateTime(1)
+	SearchAttributeFieldDateTime02 = newSearchAttributeFieldDateTime(2)
+
+	SearchAttributeFieldInt01 = newSearchAttributeFieldInt(1)
+	SearchAttributeFieldInt02 = newSearchAttributeFieldInt(2)
+
+	SearchAttributeFieldDouble01 = newSearchAttributeFieldDouble(1)
+	SearchAttributeFieldDouble02 = newSearchAttributeFieldDouble(2)
+
+	SearchAttributeFieldKeyword01 = newSearchAttributeFieldKeyword(1)
+	SearchAttributeFieldKeyword02 = newSearchAttributeFieldKeyword(2)
+	SearchAttributeFieldKeyword03 = newSearchAttributeFieldKeyword(3)
+	SearchAttributeFieldKeyword04 = newSearchAttributeFieldKeyword(4)
+
+	SearchAttributeFieldKeywordList01 = newSearchAttributeFieldKeywordList(1)
+	SearchAttributeFieldKeywordList02 = newSearchAttributeFieldKeywordList(2)
+)
+
+var (
+	_ SearchAttribute = (*searchAttributeDefinition)(nil)
+	_ SearchAttribute = (*SearchAttributeBool)(nil)
+	_ SearchAttribute = (*SearchAttributeDateTime)(nil)
+	_ SearchAttribute = (*SearchAttributeInt)(nil)
+	_ SearchAttribute = (*SearchAttributeDouble)(nil)
+	_ SearchAttribute = (*SearchAttributeKeyword)(nil)
+	_ SearchAttribute = (*SearchAttributeKeywordList)(nil)
+)
+
+type (
```

Comment:
nit: don't group all of the types together. The practice in Go is to put the type definition above its constructor followed by all of the methods.